internal/controller/argocd/api/v1alpha1/application_types.go

```diff
@@ -29,8 +31,34 @@ type ApplicationSource struct {
 	Chart          string                      `json:"chart,omitempty"`
 }
 
+// Equals compares two instances of ApplicationSource and returns true if
+// they are equal.
+func (source *ApplicationSource) Equals(other *ApplicationSource) bool {
+	if source == nil && other == nil {
+		return true
+	}
+	if source == nil || other == nil {
+		return false
+	}
+	return reflect.DeepEqual(source.DeepCopy(), other.DeepCopy())
```

Comment:
For my own edification, why the deep copies here?

Reply:
Good question, and I am not so sure myself, but I directly copied these bits from upstream.